File: mobile/android/EventNexusLiveMap/app/src/main/java/eu/eventnexus/livemap/data/repository/AuthRepository.kt
package eu.eventnexus.livemap.data.repository

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import eu.eventnexus.livemap.data.SupabaseClient
import eu.eventnexus.livemap.data.model.User
import io.github.jan.supabase.gotrue.auth
import io.github.jan.supabase.gotrue.providers.Google
import io.github.jan.supabase.gotrue.providers.builtin.Email
import io.github.jan.supabase.gotrue.providers.builtin.IDToken
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext

class AuthRepository {
    private val client = SupabaseClient.client
    private val context = SupabaseClient.getContext()
    private val analytics = AnalyticsRepository()
    
    private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "auth_prefs")
    
    private object PreferencesKeys {
        val USER_ID = stringPreferencesKey("user_id")
        val ACCESS_TOKEN = stringPreferencesKey("access_token")
    }
    
    suspend fun signIn(email: String, password: String): Result<User> = withContext(Dispatchers.IO) {
        try {
            analytics.logAuth(AnalyticsRepository.EventType.LOGIN_ATTEMPT, true, email)
            
            client.auth.signInWith(Email) {
                this.email = email
                this.password = password
            }
            
            analytics.logAuth(AnalyticsRepository.EventType.LOGIN_SUCCESS, true, email)
            handleAuthSuccess()
        } catch (e: Exception) {
            analytics.logAuth(AnalyticsRepository.EventType.LOGIN_FAILURE, false, email)
            
            // Return only the message to avoid leaking headers in UI
            val errorMessage = if (e.message?.contains("Invalid login credentials") == true) {
                "Invalid email or password"
            } else {
                e.message ?: "Login failed"
            }
            Result.failure(Exception(errorMessage))
        }
    }
    
    suspend fun signUp(email: String, password: String, name: String? = null): Result<User> = withContext(Dispatchers.IO) {
        try {
            analytics.logAuth(AnalyticsRepository.EventType.SIGNUP_ATTEMPT, true, email)
            
            client.auth.signUpWith(Email) {
                this.email = email
                this.password = password
            }
            
            analytics.logAuth(AnalyticsRepository.EventType.SIGNUP_SUCCESS, true, email)
            
            // Trigger will create profile automatically
            // Use handleAuthSuccess which will ensure profile exists
            handleAuthSuccess()
        } catch (e: Exception) {
            analytics.logAuth(AnalyticsRepository.EventType.SIGNUP_ATTEMPT, false, email)
            Result.failure(e)
        }
    }
    
    // Google Sign-In with pre-fetched ID Token
    suspend fun signInWithGoogle(idToken: String): Result<User> = withContext(Dispatchers.IO) {
        try {
            analytics.logEvent(AnalyticsRepository.EventType.LOGIN_ATTEMPT, mapOf("method" to "google"))
            
            client.auth.signInWith(IDToken) {
                this.idToken = idToken
                this.provider = Google
            }
            
            analytics.logEvent(AnalyticsRepository.EventType.LOGIN_SUCCESS, mapOf("method" to "google"))
            handleAuthSuccess()
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    private suspend fun handleAuthSuccess(): Result<User> {
        val session = client.auth.currentSessionOrNull()
        val userId = session?.user?.id ?: throw Exception("Authentication failed")
        
        context.dataStore.edit { prefs ->
            prefs[PreferencesKeys.USER_ID] = userId
            session.accessToken.let { prefs[PreferencesKeys.ACCESS_TOKEN] = it }
        }
        
        // Try to get user profile from database
        try {
            val users = client.from("users").select(Columns.ALL) {
                filter {
                    eq("id", userId)
                }
                limit(1)
            }.decodeList<User>()
            
            val user = users.firstOrNull() ?: User(
                id = userId,
                email = session.user?.email ?: "",
                name = session.user?.userMetadata?.get("full_name")?.toString() 
                    ?: session.user?.userMetadata?.get("name")?.toString()
                    ?: session.user?.email?.split("@")?.get(0) ?: "User"
            )
            
            return Result.success(user)
        } catch (e: Exception) {
            // Fallback: return basic user object from session
            return Result.success(User(
                id = userId,
                email = session.user?.email ?: "",
                name = session.user?.userMetadata?.get("full_name")?.toString()
                    ?: session.user?.userMetadata?.get("name")?.toString()
                    ?: session.user?.email?.split("@")?.get(0) ?: "User"
            ))
        }
    }
    
    suspend fun signOut(): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            analytics.logEvent(AnalyticsRepository.EventType.LOGOUT)
            client.auth.signOut()
            context.dataStore.edit { prefs -> prefs.clear() }
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun getCurrentUser(): Result<User?> = withContext(Dispatchers.IO) {
        try {
            val session = client.auth.currentSessionOrNull()
            if (session != null) {
                return@withContext handleAuthSuccess().map { it }
            }
            
            val userId = context.dataStore.data.map { it[PreferencesKeys.USER_ID] }.first()
            if (userId == null) return@withContext Result.success(null)
            
            val users = client.from("users").select(Columns.ALL) {
                filter { eq("id", userId) }
                limit(1)
            }.decodeList<User>()
            
            Result.success(users.firstOrNull())
        } catch (e: Exception) {
            Result.success(null)
        }
    }
    
    suspend fun isLoggedIn(): Boolean {
        return try {
            client.auth.currentSessionOrNull() != null
        } catch (e: Exception) {
            false
        }
    }
}

File: mobile/ios/EventNexusLiveMap/EventNexusLiveMap/Services/AuthRepository.swift
import Foundation

class AuthRepository {
    static let shared = AuthRepository()
    private let client = SupabaseManager.shared.client
    private let userDefaults = UserDefaults.standard
    
    private let userIdKey = "user_id"
    private let accessTokenKey = "access_token"
    
    private init() {}
    
    func signIn(email: String, password: String) async throws -> User {
        let session = try await client.auth.signIn(email: email, password: password)
        
        // Save session
        userDefaults.set(session.user.id.uuidString, forKey: userIdKey)
        userDefaults.set(session.accessToken, forKey: accessTokenKey)
        
        // Get user profile
        let response = try await client.database
            .from("users")
            .select()
            .eq("id", value: session.user.id.uuidString)
            .single()
            .execute()
        
        return try JSONDecoder().decode(User.self, from: response.data)
    }
    
    func signUp(email: String, password: String, name: String?) async throws -> User {
        let session = try await client.auth.signUp(email: email, password: password)
        
        let userId = session.user.id.uuidString
        
        // Create user profile
        let userData: [String: Any] = [
            "id": userId,
            "email": email,
            "name": name ?? ""
        ]
        
        try await client.database
            .from("users")
            .insert(userData)
            .execute()
        
        // Save session
        userDefaults.set(userId, forKey: userIdKey)
        userDefaults.set(session.accessToken, forKey: accessTokenKey)
        
        return User(
            id: userId,
            email: email,
            name: name,
            phone: nil,
            avatar_url: nil,
            role: "user",
            created_at: nil
        )
    }
    
    func signOut() async throws {
        try await client.auth.signOut()
        
        userDefaults.removeObject(forKey: userIdKey)
        userDefaults.removeObject(forKey: accessTokenKey)
    }
    
    func getCurrentUser() async throws -> User? {
        guard let userId = userDefaults.string(forKey: userIdKey) else {
            return nil
        }
        
        let response = try await client.database
            .from("users")
            .select()
            .eq("id", value: userId)
            .single()
            .execute()
        
        return try JSONDecoder().decode(User.self, from: response.data)
    }
    
    func isLoggedIn() -> Bool {
        return userDefaults.string(forKey: userIdKey) != nil
    }
}
